agot-bg-game-server/src/common/ingame-game-state/game-data-structure/Game.ts

```diff
@@ -148,6 +149,34 @@ export default class Game {
         return unit;
     }
 
+    transformSingle(unit: Unit, targetType: UnitType): Unit {
+        unit.region.units.delete(unit.id);
+
+        const newUnit = this.createUnit(unit.region, targetType, unit.allegiance);
+        newUnit.region.units.set(newUnit.id, newUnit);
+
+        newUnit.wounded = unit.wounded;
+
+        return newUnit
+    }
+
+    transformUnits(region: Region, units: Unit[], targetType: UnitType, entireGame: EntireGame): Unit[] {
```

Comment:
You can put these 2 methods in `IngameGameState` if they need to seend messages to clients, instead of getting an `EntireGame` in parameters.

I would have architectured this part of the code differently so that EntireGame is accessible from everywhere, but for the moment, I'd prefer to keep `Game` and the sub-classes free of any `EntireGame` and `IngameGameState`.

Reply:
No problem.
Somewhat unrelated, but I saw that each time `createUnit` is called, almost immediately after something along the lines of: `region.units.set(u.id, u)` is done. Since the unit creation is aware of the unit's region, wouldn't it be nicer to move it inside `createUnit`? (perhaps in a different branch)

Reply:
I think if I didn't do it, it is because this method would not send a message to the client, and so I didn't put `region.units.set()` inside `createUnit()` to kind of remind me to do it.

There are a lot of things I'd do differently if I had to do the project from scratch, and better function to manipulate the game state (change the order of a region, add a unit, ...) is one! 🙂

Reply:
Fair enough! We always know how to write our code better once we did it seems